tests/test_grading.py

```diff
@@ -180,3 +184,120 @@ def test_scores_with_generate(self, generate=False):
+            check_visible_scores(user, self.assignment, hidden=['Composition'], visible=['Total'])
+            check_visible_scores(user, self.assignment2, hidden=['Total'], visible=['Composition'])
+
+        # Hiding score only affect targetted assignment
+        self.login(self.staff1.email)
+        endpoint = '/admin/course/{}/assignments/{}/publish'.format(self.course.id, self.assignment.id)
+        response = self.client.post(endpoint, data={'hide':True})
+        self.assert_200(response)
+        source = response.get_data().decode('utf-8')
+        self.assertTrue("Hid {} {} scores".format(self.assignment.display_name, 'Total') in source)
+        for user in users:
+            check_visible_scores(user, self.assignment, hidden=['Total', 'Composition'])
+            check_visible_scores(user, self.assignment2, hidden=['Total'], visible=['Composition'])
+
+        self.login(self.staff1.email)
+        endpoint = '/admin/course/{}/assignments/{}/publish'.format(self.course.id, self.assignment2.id)
+        response = self.client.post(endpoint, data={})
+        self.assert_200(response)
+        for user in users:
+            check_visible_scores(user, self.assignment, hidden=['Total', 'Composition'])
+            check_visible_scores(user, self.assignment2, visible=['Composition', 'Total'])
+
+        # Cannot publish an already published grade
+        self.login(self.staff1.email)
+        endpoint = '/admin/course/{}/assignments/{}/publish'.format(self.course.id, self.assignment2.id)
+        response = self.client.post(endpoint, data={})
+        self.assert_200(response)
+        source = response.get_data().decode('utf-8')
+        self.assertTrue("{} scores for {} already published".format('Total', self.assignment2.display_name) in source)
+        for user in users:
+            check_visible_scores(user, self.assignment, hidden=['Total', 'Composition'])
+            check_visible_scores(user, self.assignment2, visible=['Composition', 'Total'])
+
+        # Cannot hide a nonpublished grade
+        self.login(self.staff1.email)
+        endpoint = '/admin/course/{}/assignments/{}/publish'.format(self.course.id, self.assignment.id)
+        response = self.client.post(endpoint, data={"hide":True})
+        self.assert_200(response)
+        source = response.get_data().decode('utf-8')
+        self.assertTrue("{} scores for {} already hidden".format('Total', self.assignment.display_name) in source)
+        for user in users:
+            check_visible_scores(user, self.assignment, hidden=['Total', 'Composition'])
+            check_visible_scores(user, self.assignment2, visible=['Composition', 'Total'])
+
+        # If assignment is not visible, cannot publish
+        self.assignment.visible = False
+        self.login(self.staff1.email)
+        endpoint = '/admin/course/{}/assignments/{}/publish'.format(self.course.id, self.assignment.id)
+        response = self.client.post(endpoint, data={})
```

Comment:
For future reference: follow_redirects will follow redirects if you want to test the final page instead